sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp

```diff
@@ -49,12 +51,109 @@ namespace Azure { namespace Core { namespace Internal { namespace Http {
+        : HttpPipeline(
+            clientOptions,
+            telemetryServiceName,
+            telemetryServiceVersion,
+            std::move(perRetryPolicies),
+            {})
+    {
+    }
+
+    /**
+     * @brief Construct a new Http Pipeline object from clientOptions.
+     *
+     * @remark The client options includes per retry and per call policies which are merged with the
+     * service-specific per retry policies.
+     *
+     * @param clientOptions The SDK client options.
+     * @param telemetryServiceName The name of the service for sending telemetry.
+     * @param telemetryServiceVersion The version of the service for sending telemetry.
+     * @param perRetryPolicies The service-specific per retry policies.
+     * @param perCallPolicies The service-specific per call policies.
+     */
+    explicit HttpPipeline(
+        ClientOptions const& clientOptions,
+        std::string const& telemetryServiceName,
+        std::string const& telemetryServiceVersion,
+        std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>>&& perRetryPolicies,
+        std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>>&& perCallPolicies)
+    {
+      auto const& perCallClientPolicies = clientOptions.GetPerCallPolicies();
+      auto const& perRetryClientPolicies = clientOptions.GerPerRetryPolicies();
+      // Adding 5 for:
+      // - TelemetryPolicy
+      // - RequestIdPolicy
+      // - RetryPolicy
+      // - LoggingPolicy
+      // - TransportPolicy
+      auto pipelineSize = perCallClientPolicies.size() + perRetryClientPolicies.size()
+          + perRetryPolicies.size() + perCallPolicies.size() + 5;
+
+      m_policies.reserve(pipelineSize);
+
+      // client-options per call policies. End-user policies execute before client's policies.
+      for (auto& policy : perCallClientPolicies)
+      {
+        m_policies.emplace_back(policy->Clone());
+      }
+      // service-specific per call policies
+      for (auto& policy : perCallPolicies)
+      {
+        m_policies.emplace_back(policy->Clone());
+      }
+
+      // Request Id
+      m_policies.emplace_back(std::make_unique<Azure::Core::Http::RequestIdPolicy>());
+      // Telemetry
+      m_policies.emplace_back(std::make_unique<Azure::Core::Http::TelemetryPolicy>(
+          telemetryServiceName, telemetryServiceVersion, clientOptions.Telemetry));
+
+      // Retry policy
+      m_policies.emplace_back(
+          std::make_unique<Azure::Core::Http::RetryPolicy>(clientOptions.Retry));
+
+      // client options per retry policies.
+      for (auto& policy : perRetryClientPolicies)
+      {
+        m_policies.emplace_back(policy->Clone());
+      }
+      // service-specific per retry policies. Services like storage have some policies which needs
+      // to execute as the last policy.
+      for (auto& policy : perRetryPolicies)
+      {
+        m_policies.emplace_back(policy->Clone());
+      }
```

Comment:
client policies should go after service policies, to enable customers to overwrite some default request headers

Reply:
need another parameter in this function for auth policy

Reply:
Done